[ENH/FIX] Re-adding database migrations (#50)
TIGRLab/dashboard: README.md

   * [Dashboard Ansible Role](#dashboard-ansible-role)  
     * [dashboard_ini_template.j2](#dashboard_ini_templatej2)
     * [dashboard.ini.j2](#dashboardinij2)
   * [Postgres Ansible Role](#postgres-ansible-role)  
     * [Postgresql Configuration](#postgresql-configuration)
     * [Backups](#backups)
6. [Code Tips](#code-tips)
   * [Jinja Templating](#jinja-templating)
   * [SQLAlchemy](#sqlalchemy)
   * [Making Database Changes](#making-database-changes)

## Technologies
The app is supported by a backend [postgreSQL](https://www.postgresql.org/)
database, hosted on the srv-postgres.camhres.ca virtual server. The front end is
programmed in python using the [Flask](http://flask.pocoo.org/) framework (Strongly recommend
[this tutorial](https://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-i-hello-world) if
you are trying to get familiar with flask) with
[SQLAlchemy](https://www.sqlalchemy.org/). The web server itself is
[NGINX](https://www.nginx.com/) with
[uWSGI](https://uwsgi-docs.readthedocs.io/en/latest/). Authentication is handled
with [OAuth](https://en.wikipedia.org/wiki/OAuth) and we currently support

> github for login.

Reviewer: ```suggestion
GitHub for login.
```

sudo -u postgres createuser -s <your_username>
```

### Setting Up the Database Tables
Once postgres is configured you can make a blank database and generate the
needed tables with:
```
createdb dashboard
flask db upgrade
```

> If you're in TIGRLab and wish to work with a pre-populated database the easiest

Reviewer: ```suggestion
If you're in TIGRLab and wish to work with a pre-populated database, the easiest
```

way is to load in an old backup of the real database (ask @DESm1th for this).
Once you have the backup you can set up your database by running the following
as a user with database creation privileges
([see this section for configuring these privileges](#configuring-postgres-for-the-dashboard)):
```
createdb dashboard
# If no host is specified psql will connect to postgres on your machine
psql dashboard -f $path_to_your_backup
```

#### Superuser Dashboard Access

> If this is a development set up you may also want to give your user account

Reviewer: ```suggestion
If this is a development setup, you may also want to give your user account
```

> superuser access to the database for convenience. To do this you must connect

Reviewer: ```suggestion
superuser access to the database for convenience. To do this, you must connect
```

to the database as a user that already has superuser permissions (on most new

> installations this would be the postgres user). Run the following in your

Reviewer: ```suggestion
installations, this would be the postgres user). Run the following in your
```

copies at all times.

## Code Tips
This is just a collection of small things to be aware of if you're going to write

> code for the dashboard. To help with profiling and debugging consider using the

Reviewer: ```suggestion
code for the dashboard. To help with profiling and debugging, consider using the
```

Reviewer: Maybe eventually, we can create a `CONTRIBUTING.md` file and move this content there.

Author: Yeah definitely!

[Flask Debug Toolbar](https://flask-debugtoolbar.readthedocs.io/en/latest/index.html)
(Thanks Mike for finding that! - Dawn). The profiler can help identify what's generating
most of the load time and the SQLAlchemy tab can let you know if your code is accidentally
generating a huge number of queries.

### Jinja Templating
Flask's HTML templates use Jinja templating to render pages
([docs here](http://jinja.pocoo.org/docs/2.10/)). The most important thing to be aware
of is that while you can break up your pages into smaller, more readable,
chunks by saving html in another file and then importing it with something like
```jinja
{% include 'my_other_file.html' %}
```
performance wise this is not always a good idea. Each and every time the 'include'

> statement is read when the website's page is loaded the included file has to be

Reviewer: ```suggestion
statement is read when the website's page is loaded, the included file has to be
```

read from the filesystem. File reads are (relatively) slow and if the include is

> inside of a loop with a large number of iterations you can easily add extra seconds

Reviewer: ```suggestion
inside of a loop with a large number of iterations, you can easily add extra seconds
```

of load time for a minimal boost in HTML readability.

Some tips to get the most out of Jinja without adding too much overhead:
  - If you have a loop and want to 'include' the body of the loop from another
    file, it's better to keep the loop inside the included file (so the file is
    opened and read once, rather than once per iteration)
  - If you have an 'if' statement and the body of it is included from another

>     snippet it's better to keep the 'if' in your original file, so you dont

Reviewer: ```suggestion
    snippet, it's better to keep the 'if' in your original file so you don't
```

for timepoint in cmh.timepoints:
   # Do some stuff with the timepoint record here
```
This loop would generate one query to the database for each timepoint that has
the site 'CMH'. If code like that were embedded in a function called in a jinja

> template then those extra queries would hit every time a user loaded the page.

Reviewer: ```suggestion
template, then those extra queries would hit every time a user loaded the page.
```
